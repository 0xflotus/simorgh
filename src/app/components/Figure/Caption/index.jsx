import React, { Fragment } from 'react';
import { string } from 'prop-types';
import styled from 'styled-components';
import {
  FF_NEWS_SANS_REG,
  C_STORM,
  GEL_SPACING,
} from '../../../lib/constants/styles';
import VisuallyHiddenText from '../../VisuallyHiddenText';
import mediaQuery from '../../../helpers/mediaQueries';

const StyledCaption = styled.figcaption`
  background-color: #d5d0cd;
  color: ${C_STORM};
  font-family: ${FF_NEWS_SANS_REG};
  padding: ${GEL_SPACING};

  // Font styling below is a subset of BBC GEL Typography "Long Primer"
  font-size: 0.9375em;
  line-height: 1.125rem;
  ${mediaQuery.laptopAndLarger} {
    font-size: 0.875em;
  }
`;

const Caption = ({ text }) => (
  <Fragment>
    <VisuallyHiddenText>Image caption, </VisuallyHiddenText>
    <StyledCaption>{text}</StyledCaption>
  </Fragment>
);

Caption.propTypes = {
  text: string.isRequired,
};

export default Caption;
