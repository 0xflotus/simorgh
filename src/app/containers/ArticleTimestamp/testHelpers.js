import moment from 'moment-timezone';

const magnitudes = {
  days: 24 * 60 * 60 * 1000,
  hours: 60 * 60 * 1000,
  minutes: 60 * 1000,
  seconds: 1000,
  milliseconds: 1,
};

const applyTimeDifference = (timestamp, timeDifference, isAdding) => {
  let acc = timestamp;

  const keyNames = Object.keys(timeDifference);

  keyNames.forEach(diff => {
    if (isAdding) {
      acc += timeDifference[diff] * magnitudes[diff];
    } else {
      acc -= timeDifference[diff] * magnitudes[diff];
    }
  });

  return acc;
};

const subtractTimeDifference = (timestamp, timeDifference) =>
  applyTimeDifference(timestamp, timeDifference, false);

const addTimeDifference = (timestamp, timeDifference) =>
  applyTimeDifference(timestamp, timeDifference, true);

export const timestampGenerator = timeDifference =>
  subtractTimeDifference(Date.now(), timeDifference);

export const multipleSameDayTimestampsGenerator = ({ intervals, date }) => {
  /**
   * This function accepts an array of intervals that represent the time
   * difference between specific moments in the past and Date.now(), and
   * returns an array of timestamps for those moments.
   *
   * The earliest one is set to 12:00am local time so that as long as all
   * intervals are less than 24 hours the timestamps returned will all be on
   * the same day, according to the machine's timezone.
   *
   * Note, the first interval provided must be the largest. If a date is not
   * provided it will use the current date.
   */

  if (!intervals || !Array.isArray(intervals)) {
    throw new Error(
      'Invalid intervals argument passed to multipleSameDayTimestampsGenerator',
    );
  }

  const firstTimestamp = moment(date)
    .startOf('day')
    .valueOf();
  const lastTimestamp = addTimeDifference(firstTimestamp, intervals[0]);
  return intervals
    .map(interval => subtractTimeDifference(lastTimestamp, interval))
    .concat(lastTimestamp);
};

export const isBritishSummerTime = timestamp =>
  moment.tz(timestamp, 'Europe/London').isDST();
