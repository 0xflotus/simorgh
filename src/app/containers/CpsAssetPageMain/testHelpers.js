import { render, waitForDomChange } from '@testing-library/react';

const renderWithHelmet = component => {
  /*
   * Similar to this problem https://github.com/testing-library/react-testing-library/issues/402
   * This helper was created to solve the problem of rendering helmet/head contents in snapshots.
   * Pass in a component that uses helmet somewhere in the component tree.
   * The full html tree is returned which you can then use to snapshot helmet/head contents.
   */

  const { container } = render(component);
  const noop = () => {};
  const ARBITRARY_TIMEOUT = 100; // seems long enough for any dom mutations to occur
  const headElement = document.querySelector('head');

  headElement.innerHTML = ''; // clear out head mutations from previous tests

  return waitForDomChange({
    container: headElement,
    timeout: ARBITRARY_TIMEOUT,
  })
    .catch(noop) // handle a waitForDomChange timeout
    .then(mutationsList => {
      const headMutationDetected = mutationsList && mutationsList.length;

      if (headMutationDetected) {
        // helmet was probably used so we should get the full html

        const htmlElement = document.querySelector('html');

        const helmetElements = document.querySelectorAll(
          '[data-react-helmet="true"]',
        );

        const removeHelmetAttributes = el =>
          el.removeAttribute('data-react-helmet'); // remove react-helmet attribute noise from elements

        removeHelmetAttributes(htmlElement); // remove react-helmet attribute noise from elements

        Array.from(helmetElements).forEach(removeHelmetAttributes);

        return { container: document.querySelector('html') };
      }

      return { container };
    });
};

export default (component, done) => {
  const removeWrappingDiv = container => container.firstChild;
  renderWithHelmet(component).then(({ container }) => {
    const hasOneChild = container.children.length === 1;
    /*
     * if the container has more than one child then it's a component that uses a
     * fragment at the top level so we should not select the first child because it
     * wouldn't snapshot the whole component
     */
    expect(
      hasOneChild ? removeWrappingDiv(container) : container,
    ).toMatchSnapshot();

    done();
  });
};
