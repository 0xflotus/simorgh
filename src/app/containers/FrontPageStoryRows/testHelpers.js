import { pathOr } from 'ramda';
import uuid from 'uuid';
import fixture from '#data/pidgin/frontpage';
import rtlFixture from '#data/urdu/frontpage';

const getFixtureData = dir => (dir === 'ltr' ? fixture : rtlFixture);

const getPromoFixtures = dir =>
  pathOr(null, ['content', 'groups'], getFixtureData(dir))
    .flatMap(group => pathOr(null, ['items'], group))
    .filter(item => pathOr(null, ['assetTypeCode'], item) === 'PRO')
    .map(item => ({ id: uuid(), ...item }));

export default getPromoFixtures;
