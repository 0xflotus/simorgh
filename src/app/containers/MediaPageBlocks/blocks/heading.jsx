import React from 'react';
import { Headline } from '@bbc/psammead-headings';

// eslint-disable-next-line react/prop-types
const HeadingContainer = ({ uuid, script, service, idAttr, text }) => {
  if (!text) return null;

  return (
    <Headline key={uuid} script={script} service={service} id={idAttr}>
      {text}
    </Headline>
  );
};

export default HeadingContainer;
