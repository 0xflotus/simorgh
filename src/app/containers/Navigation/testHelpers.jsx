import React from 'react';

export const dropdownTestId = 'dropdown';
export const scrollableTestId = 'scrollable-list';

export const scrollableListItems = (
  <ul data-testid={scrollableTestId}>
    <li>List Items</li>
  </ul>
);

export const dropdownListItems = (
  <div data-testid={dropdownTestId}>Dropdown</div>
);
