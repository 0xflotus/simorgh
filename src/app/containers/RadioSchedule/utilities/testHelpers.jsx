import React from 'react';
import RadioScheduleContainer from '..';
import { RequestContextProvider } from '#contexts/RequestContext';
import { ServiceContextProvider } from '#contexts/ServiceContext';
import { ToggleContext } from '#contexts/ToggleContext';

const localRadioScheduleEndpoint = service => {
  const localhostURL = 'http://localhost:7080';
  const localServiceURL = `${localhostURL}/${service}`;
  return `${localServiceURL}/bbc_${service}_radio/schedule.json`;
};

const getToggleState = enabled => ({
  local: { radioSchedule: { enabled } },
  test: { radioSchedule: { enabled } },
});

/* eslint-disable react/prop-types */
const RadioSchedulesWithContext = ({
  service,
  radioScheduleToggle = false,
}) => (
  <ToggleContext.Provider
    value={{ toggleState: getToggleState(radioScheduleToggle) }}
  >
    <RequestContextProvider
      isAmp={false}
      pageType="frontPage"
      service={service}
      pathname={`/${service}`}
    >
      <ServiceContextProvider service={service}>
        <RadioScheduleContainer
          endpoint={localRadioScheduleEndpoint(service)}
        />
      </ServiceContextProvider>
    </RequestContextProvider>
  </ToggleContext.Provider>
);
/* eslint-enable react/prop-types */

export default RadioSchedulesWithContext;
