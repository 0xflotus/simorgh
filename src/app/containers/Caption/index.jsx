import React, { useContext } from 'react';
import { objectOf, any, string } from 'prop-types';
import VisuallyHiddenText from '@bbc/psammead-visually-hidden-text';
import Caption from '@bbc/psammead-caption';
import Paragraph from '@bbc/psammead-paragraph';
import deepGet from '../../helpers/json/deepGet';
import { ServiceContext } from '../../contexts/ServiceContext';
import Blocks from '../Blocks';
import Fragment from '../Fragment';
import InlineLink from '../InlineLink';

const componentsToRender = { fragment: Fragment, urlLink: InlineLink };

const chooseOffscreenText = (mediaType, videoCaption, imageCaption) => {
  let offscreenText = 'Caption, ';
  if (mediaType === 'video') {
    offscreenText = videoCaption;
  }
  if (mediaType === 'image') {
    offscreenText = imageCaption;
  }

  return offscreenText;
};
const renderParagraph = paragraphBlock => (
  <Paragraph key={deepGet([0, 'model', 'text'], paragraphBlock)}>
    <Blocks blocks={paragraphBlock} componentsToRender={componentsToRender} />
  </Paragraph>
);

const renderCaption = (paragraphBlocks, offscreenText, script) => (
  <Caption script={script}>
    {offscreenText ? (
      <VisuallyHiddenText>{offscreenText}</VisuallyHiddenText>
    ) : null}
    {paragraphBlocks.map(block => {
      const paragraphBlock = deepGet(['model', 'blocks'], block);
      return renderParagraph(paragraphBlock);
    })}
  </Caption>
);

const CaptionContainer = ({ block, type }) => {
  const {
    script,
    imageCaptionOffscreenText,
    videoCaptionOffscreenText,
  } = useContext(ServiceContext);
  const offscreenText = chooseOffscreenText(
    type,
    videoCaptionOffscreenText,
    imageCaptionOffscreenText,
  );
  const paragraphBlocks = deepGet(
    ['model', 'blocks', 0, 'model', 'blocks'],
    block,
  );

  return renderCaption(paragraphBlocks, offscreenText, script);
};

CaptionContainer.propTypes = {
  block: objectOf(any).isRequired,
  type: string,
};

CaptionContainer.defaultProps = {
  type: '',
};

export default CaptionContainer;
