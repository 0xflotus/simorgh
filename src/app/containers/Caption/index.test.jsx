import React from 'react';
import Caption from '.';
import { shouldMatchSnapshot } from '../../helpers/tests/testHelpers';
import { ServiceContext } from '../../contexts/ServiceContext';
import { blockContainingText } from '../../models/blocks';

const newsServiceContextStub = {
  imageCaptionOffscreenText: 'Image caption, ',
};
const persianServiceContextStub = {
  imageCaptionOffscreenText: ' ، عنوان تصویر',
};

const captionBlock = blockContainingText('caption', 'Some caption text...');

const captionBlock3Paragraphs = {
  model: {
    blocks: [
      {
        model: {
          blocks: [
            {
              model: {
                blocks: [
                  {
                    model: {
                      attributes: [],
                      text: 'This is paragraph 1',
                    },
                    type: 'fragment',
                  },
                ],
                text: 'This is paragraph 1',
              },
              type: 'paragraph',
            },
            {
              model: {
                blocks: [
                  {
                    model: {
                      attributes: [],
                      text: 'This is paragraph 2',
                    },
                    type: 'fragment',
                  },
                ],
                text: 'This is paragraph 2',
              },
              type: 'paragraph',
            },
            {
              model: {
                blocks: [
                  {
                    model: {
                      attributes: [],
                      text: 'Paragraph 3',
                    },
                    type: 'fragment',
                  },
                ],
                text: 'Paragraph 3',
              },
              type: 'paragraph',
            },
          ],
        },
        type: 'text',
      },
    ],
  },
  type: 'caption',
};

const CaptionWithContext = (block, contextStub) => (
  <ServiceContext.Provider value={contextStub}>
    <Caption block={block} />
  </ServiceContext.Provider>
);

shouldMatchSnapshot(
  'should render caption text with example News offscreen text',
  CaptionWithContext(captionBlock, newsServiceContextStub),
);

shouldMatchSnapshot(
  'should render caption text with example Farsi offscreen text',
  CaptionWithContext(captionBlock, persianServiceContextStub),
);

shouldMatchSnapshot(
  'should render caption text with no VisuallyHiddenText component when no imageCaptionOffscreenText is defined in ServiceContext',
  CaptionWithContext(captionBlock, { imageCaptionOffscreenText: undefined }),
);

shouldMatchSnapshot(
  'should render caption with mutiple paragraphs',
  CaptionWithContext(captionBlock3Paragraphs, newsServiceContextStub),
);
