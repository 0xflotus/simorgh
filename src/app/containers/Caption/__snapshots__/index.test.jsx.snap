// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render caption text with example Farsi offscreen text 1`] = `
.c1 {
  -webkit-clip-path: inset(100%);
  clip-path: inset(100%);
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  height: 1px;
  overflow: hidden;
  position: absolute;
  width: 1px;
}

.c0 {
  font-size: 0.9375rem;
  line-height: 1.125rem;
  color: #757575;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-style: italic;
  padding: 0.5rem 0.5rem;
  width: 100%;
}

.c2 {
  color: #3F3F42;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  padding-bottom: 1.5rem;
  margin: 0;
  font-size: 0.9375rem;
  line-height: 1.25rem;
}

@media (min-width:37.5em) {
  .c0 {
    font-size: 0.875rem;
  }
}

@media (min-width:25em) and (max-width:62.9375em) {
  .c0 {
    padding: 0.5rem 1rem;
  }
}

@media (min-width:63em) {
  .c0 {
    padding: 0.5rem 0;
  }
}

@media (min-width:20rem) {
  .c2 {
    font-size: 1rem;
    line-height: 1.375rem;
  }
}

<figcaption
  className="c0"
>
  <span
    className="c1"
  >
     ، عنوان تصویر
  </span>
  <p
    className="c2"
  >
    Some caption text...
  </p>
</figcaption>
`;

exports[`should render caption text with example News offscreen text 1`] = `
.c1 {
  -webkit-clip-path: inset(100%);
  clip-path: inset(100%);
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  height: 1px;
  overflow: hidden;
  position: absolute;
  width: 1px;
}

.c0 {
  font-size: 0.9375rem;
  line-height: 1.125rem;
  color: #757575;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-style: italic;
  padding: 0.5rem 0.5rem;
  width: 100%;
}

.c2 {
  color: #3F3F42;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  padding-bottom: 1.5rem;
  margin: 0;
  font-size: 0.9375rem;
  line-height: 1.25rem;
}

@media (min-width:37.5em) {
  .c0 {
    font-size: 0.875rem;
  }
}

@media (min-width:25em) and (max-width:62.9375em) {
  .c0 {
    padding: 0.5rem 1rem;
  }
}

@media (min-width:63em) {
  .c0 {
    padding: 0.5rem 0;
  }
}

@media (min-width:20rem) {
  .c2 {
    font-size: 1rem;
    line-height: 1.375rem;
  }
}

<figcaption
  className="c0"
>
  <span
    className="c1"
  >
    Image caption, 
  </span>
  <p
    className="c2"
  >
    Some caption text...
  </p>
</figcaption>
`;

exports[`should render caption text with no VisuallyHiddenText component when no imageCaptionOffscreenText is defined in ServiceContext 1`] = `
.c0 {
  font-size: 0.9375rem;
  line-height: 1.125rem;
  color: #757575;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-style: italic;
  padding: 0.5rem 0.5rem;
  width: 100%;
}

.c1 {
  color: #3F3F42;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  padding-bottom: 1.5rem;
  margin: 0;
  font-size: 0.9375rem;
  line-height: 1.25rem;
}

@media (min-width:37.5em) {
  .c0 {
    font-size: 0.875rem;
  }
}

@media (min-width:25em) and (max-width:62.9375em) {
  .c0 {
    padding: 0.5rem 1rem;
  }
}

@media (min-width:63em) {
  .c0 {
    padding: 0.5rem 0;
  }
}

@media (min-width:20rem) {
  .c1 {
    font-size: 1rem;
    line-height: 1.375rem;
  }
}

<figcaption
  className="c0"
>
  <p
    className="c1"
  >
    Some caption text...
  </p>
</figcaption>
`;

exports[`should render caption with mutiple paragraphs 1`] = `
.c1 {
  -webkit-clip-path: inset(100%);
  clip-path: inset(100%);
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  height: 1px;
  overflow: hidden;
  position: absolute;
  width: 1px;
}

.c0 {
  font-size: 0.9375rem;
  line-height: 1.125rem;
  color: #757575;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-style: italic;
  padding: 0.5rem 0.5rem;
  width: 100%;
}

.c2 {
  color: #3F3F42;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  padding-bottom: 1.5rem;
  margin: 0;
  font-size: 0.9375rem;
  line-height: 1.25rem;
}

@media (min-width:37.5em) {
  .c0 {
    font-size: 0.875rem;
  }
}

@media (min-width:25em) and (max-width:62.9375em) {
  .c0 {
    padding: 0.5rem 1rem;
  }
}

@media (min-width:63em) {
  .c0 {
    padding: 0.5rem 0;
  }
}

@media (min-width:20rem) {
  .c2 {
    font-size: 1rem;
    line-height: 1.375rem;
  }
}

<figcaption
  className="c0"
>
  <span
    className="c1"
  >
    Image caption, 
  </span>
  <p
    className="c2"
  >
    This is paragraph 1
  </p>
  <p
    className="c2"
  >
    This is paragraph 2
  </p>
  <p
    className="c2"
  >
    Paragraph 3
  </p>
</figcaption>
`;
